core/include/scipp/core/dtype.h

```diff
@@ -14,12 +12,71 @@
+// for two different types T1 and T2, but we currently do not have a better
+// solution.
+// force compiler error if not specialized
+template <class T> constexpr DType dtype{T::missing_specialization_of_dtype};
+// basics types start at 0
+template <> constexpr DType dtype<void>{0};
+template <> constexpr DType dtype<double>{1};
+template <> constexpr DType dtype<float>{2};
+template <> constexpr DType dtype<int64_t>{3};
+template <> constexpr DType dtype<int32_t>{4};
+template <> constexpr DType dtype<bool>{5};
+template <> constexpr DType dtype<std::string>{6};
+template <> constexpr DType dtype<time_point>{7};
+template <> constexpr DType dtype<Eigen::Vector3d>{8};
+template <> constexpr DType dtype<Eigen::Matrix3d>{9};
+class SubbinSizes;
+template <> constexpr DType dtype<SubbinSizes>{10};
+// span<T> start at 100
+template <> constexpr DType dtype<span<const double>>{100};
+template <> constexpr DType dtype<span<const float>>{101};
+template <> constexpr DType dtype<span<const int64_t>>{102};
+template <> constexpr DType dtype<span<const int32_t>>{103};
+template <> constexpr DType dtype<span<const bool>>{104};
+template <> constexpr DType dtype<span<const std::string>>{105};
+template <> constexpr DType dtype<span<const time_point>>{106};
+template <> constexpr DType dtype<span<const Eigen::Vector3d>>{107};
+// span<const T> start at 200
+template <> constexpr DType dtype<span<double>>{200};
+template <> constexpr DType dtype<span<float>>{201};
+template <> constexpr DType dtype<span<int64_t>>{202};
+template <> constexpr DType dtype<span<int32_t>>{203};
+template <> constexpr DType dtype<span<bool>>{204};
+template <> constexpr DType dtype<span<std::string>>{205};
+template <> constexpr DType dtype<span<time_point>>{206};
+template <> constexpr DType dtype<span<Eigen::Vector3d>>{207};
+// std containers start at 300
+template <> constexpr DType dtype<std::pair<int32_t, int32_t>>{300};
+template <> constexpr DType dtype<std::pair<int64_t, int64_t>>{301};
+template <> constexpr DType dtype<std::unordered_map<double, int64_t>>{302};
+template <> constexpr DType dtype<std::unordered_map<double, int32_t>>{303};
+template <> constexpr DType dtype<std::unordered_map<float, int64_t>>{304};
+template <> constexpr DType dtype<std::unordered_map<float, int32_t>>{305};
+template <> constexpr DType dtype<std::unordered_map<int64_t, int64_t>>{306};
+template <> constexpr DType dtype<std::unordered_map<int64_t, int32_t>>{307};
+template <> constexpr DType dtype<std::unordered_map<int32_t, int64_t>>{308};
+template <> constexpr DType dtype<std::unordered_map<int32_t, int32_t>>{309};
+template <> constexpr DType dtype<std::unordered_map<bool, int64_t>>{310};
+template <> constexpr DType dtype<std::unordered_map<bool, int32_t>>{311};
+template <>
+constexpr DType dtype<std::unordered_map<std::string, int64_t>>{312};
+template <>
+constexpr DType dtype<std::unordered_map<std::string, int32_t>>{313};
```

Comment:
Not fully complete - however, could this be done without the long hand-rolled set of specialisations? For example. Could we define all supported types in a tuple and use the index as the id.

```
template <class T, class Tuple>
struct Index;

template <class T, class... Types>
struct Index<T, std::tuple<T, Types...>> {
    static const int32_t value = 0;
};

template <class T, class U, class... Types>
struct Index<T, std::tuple<U, Types...>> {
    static const int32_t value = 1 + Index<T, std::tuple<Types...>>::value;
};

```

Reply:
Or generate it with CMake? This would also allow us to simplify nested types like `span<int>`.

Reply:
I vaguely remember we had something like Owen's suggestion before moving to `std::type_index`. We would still need to have offsets to handle instantiations from other modules. I think I'll leave it for now until we a sure this is really the right solution.

Reply:
Yes. There are enough changes here already.